### ISSUE TITLE
zip without compression extracts to empty files

### ISSUE THREAD

**Issue by ericcj** [rubyzip/rubyzip#236]
i created a zip using https://github.com/pixelglow/zipzap with compress:NO on the entries and it unzips fine using the normal unzip command line utility (Info-ZIP) and Archive Utility on my os x laptop.  however, when i try to extract it with rubyzip it creates zero-byte files for all the entries.  here's what one of the entries and its input stream looks like.  sysread on that stream always returns empty string. 

```
#<Zip::Entry:0x007f9abcab60e0 @local_header_offset=99599183, @local_header_size=nil, @internal_file_attributes=0, @external_file_attributes=2175008768, @header_signature=33639248, @version_needed_to_extract=10, @version=30, @ftype=:file, @filepath=nil, @gp_flags=2056, @follow_symlinks=false, @restore_times=true, @restore_permissions=false, @restore_ownership=false, @unix_uid=nil, @unix_gid=nil, @unix_perms=420, @dirty=false, @fstype=3, @zipfile="/Users/ej/17ba3bfa-139f-11e5-9c00-2a828cf49bc3.zip", @name="C45664C1-ACC5-413E-9534-C7BF05DC5338.jpg", @comment="", @extra={}, @compressed_size=2239128, @crc=375996240, @compression_method=0, @size=2239128, @time=2015-06-15 16:42:50 -0400, @last_mod_time=34137, @last_mod_date=18127, @name_length=40, @extra_length=0, @comment_length=0>

#<Zip::InputStream:0x007f9ac2549a98 @lineno=0, @pos=0, @output_buffer="", @archive_io=#<File:/Users/ej/17ba3bfa-139f-11e5-9c00-2a828cf49bc3.zip>, @decompressor=#<Zip::PassThruDecompressor:0x007f9ac25495e8 @input_stream=#<File:/Users/ej/17ba3bfa-139f-11e5-9c00-2a828cf49bc3.zip>, @chars_to_read=0, @read_so_far=0, @has_returned_empty_string=false>, @decrypter=#<Zip::NullDecrypter:0x007f9ac25499d0>, @current_entry=#<Zip::Entry:0x007f9ac25499a8 @local_header_offset=99599183, @local_header_size=70, @internal_file_attributes=1, @external_file_attributes=0, @header_signature=67324752, @version_needed_to_extract=20, @version=10, @ftype=:file, @filepath=nil, @gp_flags=2056, @follow_symlinks=false, @restore_times=true, @restore_permissions=false, @restore_ownership=false, @unix_uid=nil, @unix_gid=nil, @unix_perms=nil, @dirty=false, @fstype=0, @zipfile=#<File:/Users/ej/17ba3bfa-139f-11e5-9c00-2a828cf49bc3.zip>, @name="C45664C1-ACC5-413E-9534-C7BF05DC5338.jpg", @comment="", @extra={}, @compressed_size=0, @crc=0, @compression_method=0, @size=0, @time=2015-06-15 16:42:50 -0400, @last_mod_time=34137, @last_mod_date=18127, @name_length=40, @extra_length=0>>
```

**Comment by simonoff** [comment]
Need to look on zip file structure what this library making.

**Comment by jlhonora** [comment]
I'm facing the same issue. Basically the problem is that in lib/zip/input_stream.rb:147

```
when @current_entry.compression_method == ::Zip::Entry::STORED
   ::Zip::PassThruDecompressor.new(@archive_io, @current_entry.size)
```

`@current_entry.size == 0`. When unpacking the values from the buffer, the size is indeed zero, lib/zip/entry.rb:

```
def unpack_local_entry(buf)
  puts "unpack local entry buf: #{buf.bytes.to_a.map {|c| '%02x' % c}.join(' ')}"
```

prints `50 4b 03 04 0a 00 08 08 00 00 96 bb fe 46 00 00 00 00 00 00 00 00 00 00 00 00 0f 00 00 00`, so effectively `size = 0` (and `compressed_size = 0`).

I'm using [this zip](https://github.com/jlhonora/rubyzip/blob/master/test/data/zipzap.zip?raw=true) to replicate it. The test case is:

```
  def test_extract_zipzap
    filename = 'test/data/zipzap.zip'
    ::Zip::File.open(filename) {
      |zf|
      assert_operator zf.read('evaluation.json').size, :>, 0
    }
  end
```

Which of course fails.

I've been trying to follow rubyzip's code in order to fix it, but this is my first time dealing with zip formats and the like. Any help is appreciated.

**Comment by pharmazone** [comment]
Is there any updates or solutions?
thnx

**Comment by herdst** [comment]
I'm having the same issue. Does anyone have a solution? Thanks

**Comment by rlgomes** [comment]
Any update here ? I have zip files with NO_COMPRESSION and basically rubzip says there are no entries when uncompressing

**Comment by ameyer-pivotal** [comment]
@ericcj Take a look at PR #358. It looks like your general purpose bit 3 is set (2056 & 8 != 0), and this may be a fix for you too.

**Comment by jdleesmiller** [comment]
(I'm helping Alex with maintenance.) I see a 👍 and no further reports since #358 landed in 1.2.2, so I will close the ticket. Please comment if it is still a problem.

Thanks for the fix @ameyer-pivotal and @idoru.
